File: src/main/java/com/innercircle/surveryproject/modules/controller/SurveyAnswerController.java
package com.innercircle.surveryproject.modules.controller;

import com.innercircle.surveryproject.modules.dto.SurveyAnswerCreateDto;
import com.innercircle.surveryproject.modules.dto.SurveyAnswerDto;
import com.innercircle.surveryproject.modules.dto.SurveyAnswerKeywordDto;
import com.innercircle.surveryproject.modules.dto.SurveyAnswerResponseDto;
import com.innercircle.surveryproject.modules.global.ResponseUtils;
import com.innercircle.surveryproject.modules.service.SurveyAnswerService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.responses.ApiResponse;
import io.swagger.v3.oas.annotations.responses.ApiResponses;
import io.swagger.v3.oas.annotations.tags.Tag;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;

/**
 * 설문조사 응답 api
 */
@Tag(name = "설문조사 응답 API")
@RestController
@RequiredArgsConstructor
@RequestMapping("/api/surveyAnswer")
public class SurveyAnswerController {

    private final SurveyAnswerService surveyAnswerService;

    @Operation(summary = "설문조사 응답 제출")
    @ApiResponses(value = {
        @ApiResponse(responseCode = "201", description = "설문조사 응답 제출에 성공하였습니다."),
        @ApiResponse(responseCode = "400", description = "설문조사 응답 제출에 실패하였습니다.")
    })
    @PostMapping
    public ResponseEntity createSurveyAnswer(@RequestBody SurveyAnswerCreateDto surveyAnswerCreateDto) {

        SurveyAnswerDto surveyAnswerDto = surveyAnswerService.createSurveyAnswer(surveyAnswerCreateDto);

        return ResponseUtils.created(surveyAnswerDto, "설문조사 응답이 완료되었습니다.");
    }

    @Operation(summary = "설문조사 응답 조회 조건")
    @ApiResponses(value = {
        @ApiResponse(responseCode = "200", description = "설문조사 응답 키워드 조회에 성공하였습니다."),
        @ApiResponse(responseCode = "400", description = "설문조사 응답 키워드 조회에 실패하였습니다.")
    })
    @GetMapping("/{surveyAnswerId}/keywords")
    public ResponseEntity retrieveSurveyAnswerKeyword(@PathVariable Long surveyAnswerId) {

        List<SurveyAnswerKeywordDto> surveyAnswerKeywordDtoList = surveyAnswerService.retrieveSurveyAnswerKeywords(surveyAnswerId);

        return ResponseUtils.ok(surveyAnswerKeywordDtoList, "설문조사 응답 키워드 조회에 성공하였습니다.");
    }

    @Operation(summary = "설문조사 응답 조회")
    @ApiResponses(value = {
        @ApiResponse(responseCode = "200", description = "설문조사 응답 조회에 성공하였습니다."),
        @ApiResponse(responseCode = "400", description = "설문조사 응답 조회에 실패하였습니다.")
    })
    @GetMapping("/{surveyAnswerId}")
    public ResponseEntity retrieveSurveyAnswer(@PathVariable Long surveyAnswerId,
                                               @RequestParam(required = false) Long surveyItemId,
                                               @RequestParam(required = false) String surveyItemContent) {

        List<SurveyAnswerResponseDto> surveyAnswerDto =
            surveyAnswerService.retrieveSurveyAnswer(surveyAnswerId, surveyItemId, surveyItemContent);

        return ResponseUtils.ok(surveyAnswerDto, "설문조사 응답 조회에 성공하였습니다.");
    }

}

File: survey-api/src/main/kotlin/com/ic/surveyapi/answer/controller/SurveyAnswerController.kt
package com.ic.surveyapi.answer.controller

import com.ic.surveyapi.answer.controller.dto.SurveyAnswerRequest
import com.ic.surveyapi.answer.controller.dto.SurveyFormAnswerResponse
import com.ic.surveyapi.answer.service.SurveyAnswerService
import com.ic.surveyapi.answer.service.dto.SurveyAnswerDto
import com.ic.surveyapi.util.ObjectMapperUtil
import org.springframework.http.HttpStatus
import org.springframework.validation.annotation.Validated
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.bind.annotation.RestController
import survey.common.ApiEndpointVersionPrefix

@RestController
@RequestMapping(ApiEndpointVersionPrefix.V1_API_SURVEY_PREFIX)
class SurveyAnswerController(
    private val surveyAnswerService: SurveyAnswerService,
    private val objectMapperUtil: ObjectMapperUtil,
) {
    // TODO : URL Path 들 다시 고민해보자 .... !
    @ResponseStatus(HttpStatus.CREATED)
    @PostMapping("/{surveyFormId}/answers")
    fun submitSurveyAnswer(
        @RequestBody @Validated surveyAnswerRequest: SurveyAnswerRequest,
        @PathVariable(name = "surveyFormId", required = true) surveyFormId: String,
    ) = run { objectMapperUtil.convertClass(value = surveyAnswerRequest, clazz = SurveyAnswerDto::class.java) }
        .let { surveyAnswerService.submitSurveyAnswer(surveyFormId = surveyFormId, surveyAnswer = it) }

    // 해당 특정 설문 조사를 조회
    @GetMapping("/{surveyFormId}/answers")
    fun getSurveyAnswers(
        @PathVariable(name = "surveyFormId", required = true) surveyFormId: String,
    ): SurveyFormAnswerResponse =
        run { surveyAnswerService.getSurveyAnswerBySurveyFormId(surveyFormId = surveyFormId) }
            .let { objectMapperUtil.convertClass(value = it, clazz = SurveyFormAnswerResponse::class.java) }

    // 설문 조사 Title 기반으로 설문 조사를 조회
    @GetMapping("/answers")
    fun getSurveyAnswersBySurveyTitle(
        @RequestParam(name = "surveyTitle", required = true
    ) surveyTitle: String) =
        let { surveyAnswerService.getSurveyAnswerByTitle(surveyTitle) }

    // TODO - 응답 항목의 이름과 응답 값을 기반으로 검색 가능하도록 구현 필요

}
